### DISCUSSION TITLE
How to generate embeddings using Microsoft.ML.OnnxRuntimeGenAI with phi-3 c#

### DISCUSSION THREAD

**Discussion by pavinan** [microsoft/onnxruntime-genai#474] [Q&A]
How to generate embeddings using Microsoft.ML.OnnxRuntimeGenAI with phi-3 c#

**Comment by Ben-Epstein** [comment]
I was wondering the same!

**Reply by MaxAkbar** [reply]
Based on this it does not support embedding: [Phi3Family.md](https://github.com/microsoft/Phi-3CookBook/blob/main/md/01.Introduce/Phi3Family.md#example-of-phi-3-tasks). That said I have used [Semantic Kernel](BertOnnxTextEmbeddingGenerationService) having a local BERT model. This link is in c# but Semantic Kernel supports python and Java as well.

**Comment by rhenry74** [comment]
any way to get embeddings would be cool

**Reply by svsnux** [reply]
Is there an api for this?  This can be usefull for RAG

**Reply by rhenry74** [reply]
I have had some success using onnx to get embeddings using WhereIsAI/UAE-Large-V1
[on hugging face](https://huggingface.co/WhereIsAI/UAE-Large-V1) <- note the onnx subfolder

Maybe this onnx model could be run thru the genai tool and then utilized to get embeddings via onnxruntime-genai.
[i think this is the tooling to use](https://onnxruntime.ai/docs/genai/howto/build-model.html)

If you see a file called genai_config.json come out you have a good chance of getting it to work... i think :-)

**Comment by kunal-vaishnavi** [comment]
Can you clarify which specific information you are looking for?

1. [Weights and biases in `self.model.embed_tokens`](https://github.com/huggingface/transformers/blob/bc36c26fa62fce8b89b13d42e2dbdfafeeb102ca/src/transformers/models/llama/modeling_llama.py#L1003)
2. [`inputs_embeds`](https://github.com/huggingface/transformers/blob/bc36c26fa62fce8b89b13d42e2dbdfafeeb102ca/src/transformers/models/llama/modeling_llama.py#L826)
3. [Last hidden state before language modeling head](https://github.com/huggingface/transformers/blob/bc36c26fa62fce8b89b13d42e2dbdfafeeb102ca/src/transformers/models/llama/modeling_llama.py#L1082)
4. [Weights and biases in `self.lm_head`](https://github.com/huggingface/transformers/blob/bc36c26fa62fce8b89b13d42e2dbdfafeeb102ca/src/transformers/models/llama/modeling_llama.py#L1009)

**Reply by svsnux** [reply]
The idea is to create [BertOnnxTextEmbeddingGenerationService.cs](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/Connectors/Connectors.Onnx/BertOnnxTextEmbeddingGenerationService.cs) for Semantic Kernel / Kernel memory but for Phi3 onnx model.  

Also see [LLamaEmbedder.cs](https://github.com/SciSharp/LLamaSharp/blob/master/LLama/LLamaEmbedder.cs) from LLamaSharp.

I'm not exactly sure but I think both 1 and 3 will be helpful.

According to this python code it seems like the Last Hidden State [OnnxEmbeddingModels](https://github.com/taylorai/onnx_embedding_models/blob/main/src/onnx_embedding_models/embedding.py)

**Comment by pavinan** [comment]
It seems my misunderstanding, gen ai models are different from embeddings model. Phi3 is gen ai model. In demos I noticed Microsoft is using "smart components" local embedding model

**Comment by elephantpanda** [comment]
Personally I would like to get the output vectors for a given token position. So I can examine all the probabilities and make my own choice of which token to pick or even display the probabilites. So I can design a custom function in order to have more control over which token it pick. I would like to get the probability vectors for then entire output.

In the vision model, I would also like to be able to examine the input generated after feeding in the images and modify it if I so please.

**Comment by kunal-vaishnavi** [comment]
The embeddings (a.k.a. last hidden states) are [now accessible](https://github.com/microsoft/onnxruntime-genai/pull/1127) via ONNX Runtime GenAI. You can create the ONNX model to output the hidden states using [ONNX Runtime GenAI's model builder](https://github.com/microsoft/onnxruntime-genai/blob/main/src/python/py/models/README.md#include-last-hidden-states-output) with `--extra_options include_hidden_states=true`.
